src/binocle/core/binocle_ecs.c

```diff
@@ -35,6 +35,7 @@ bool binocle_sparse_integer_set_insert(binocle_sparse_integer_set_t *set, uint64
       new_sparse = realloc(set->sparse, new_capacity * sizeof(uint64_t));
     }
     if (new_sparse == NULL) {
+      free(new_dense);
```

Comment:
This would actually break things as the dense set would have been either allocated or reallocated when the code gets to this line. Freeing it would leave it in an inconsistent state only because the sparse set couldn't be (re)allocated. 
A better option would probably be to check if both the sets can be (re)allocated and only then copy the values over or leave them both as they are